# Question 1

Given two strings `s` and `t`, determine whether some anagram of `t` is a substring
of `s`. For example: if `s = "udacity"` and `t = "ad"`, then the function returns
True. Your function definition should look like: `question1(s, t)` and return a
boolean `True` or `False`.


In [5]:
# Turn on/off to enable/disable debugging
debug = False


def is_anagram(s, t):
    """
    # Find if strings are anagram. t anagram of s
    # @param {string, string} input strings
    # @return bool if strings are anagram of each other or not
    """
    s = list(s)
    # Sort a string and then compare with each other
    s.sort()   # Quick sort O(n*log(n))
    return s == t


def question1(s,t):
    """
    # Find out sorted(possible substring of s) and compare with sorted(t)
    # @param main_string, ana_string input strings
    # @return bool Question1 answer
    """
    global debug
    sort_t = list(t)
    sort_t.sort()     # Quick sort O(n*log(n))

    for i in range(len(s) - len(t) + 1):
        if debug:
            print (s[i: i+len(t)], t)
        if is_anagram(s[i: i+len(t)], sort_t):
            return True
    return False

# Main program
def main():
    unit_tests = [("udacity", "ad"),
                 ("udacity", "da"),
                 ("udacity", "au"),
                 ("udacity", "ciy"),
                 ("udacity", "xyz"),
                 ("udacity", ""),
                 ("udacity", "12863742302323232746"),
                 ("ad", "udacity" ),
                 ("panama", "1pana"),
                 ("banana", "anana")]

    for test in unit_tests:
        print("{} -> {}".format(test,question1(test[0],test[1])))


if __name__ == '__main__':
    main()

('udacity', 'ad') -> True
('udacity', 'da') -> True
('udacity', 'au') -> False
('udacity', 'ciy') -> False
('udacity', 'xyz') -> False
('udacity', '') -> True
('udacity', '12863742302323232746') -> False
('ad', 'udacity') -> False
('panama', '1pana') -> False
('banana', 'anana') -> True


## Answer

So the code was chosen to be the most optimal for the problem in hands. It has complexity of O(Nlogm) since it only iterates through the entry string once. There are two sort functions applied, once to the string t, which if it is large, will take some time, this has complexity O(NlogN). But in most cases string t will be short and not cause much problems. Also, there is a need for a second sort function on the substring from main string. This will have the same length as the string t, so sorting can take some time. This sort will also occur several times throughout the routine. Se if S is length 8 and t is length 2, this sort will occur 4 times total.

Sort will occur on string t with complexity O(mlogm), m being the length of string t.
Sort will occur on substring of S with complexity O(nlogm), n being length of string S.

Given the test cases bellow:

* Case 1: ("udacity", "ad") -> True
string with 2 characters reversed from order in S

*  Case 2: ("udacity", "da") -> True
string with 2 characters same as in order in S

*  Case 3: ("udacity", "au") -> False
string with 2 alternative characters and reversed from order in S

*  Case 4: ("udacity, dacity") -> True
string with 6 characters same as in order in S

*  Case 5: ("udacity", "ciy") -> False
string with 2 characters same as in order in S and 1 character out of order.

*  Case 6: ("udacity", "xyz") -> False
string with new characters not in string S.

* Case 7: ("udacity", "") -> True
empty string.

*  Case 8: ("ad", "udacity" ) -> False
string t's length is greater than string S.

*  Case 9: ("panama", "1pana" ) -> False
string t's contains a number before.

*  Case 10: ("banana", "anana" ) -> False
string t's is an anagram of string S

All tests are passed as it can be seen above.


# Question 2
Given a string a, find the longest palindromic substring contained in a.
Your function definition should look like `question2(a)`, and return a string.


In [30]:
debug = False

def isPalindrome(string):
    """
    # @param string s input string
    # @return bool if string is palindrome or not
    """
    if not string:
        return False
    # reverse compare
    return string == string[::-1]


def question2(string):
    """
    # @param string s input string
    # @return string the longest palindromic substring
    """
    global debug
    
    if not string: # if null entry
        return ""

    n = len(string)
    longest, left, right = 0, 0, 0
    for i in range(0, n):
        for j in range(i + 1, n + 1):
            substr = string[i:j] # going through all substrings possible
            
            
            if isPalindrome(substr) and len(substr) > longest: # see if it is palindrome and is longer than the one before
                longest = len(substr)
                left, right = i, j
    # construct longest substr
    result = string[left:right]
    return result

# Main program
def main():
    unit_tests = ["abuttuba",
                 "acaramanamaraca",
                 "adogapanicinapagoda",
                 "anutforajaroftuna",
                 "asantaatnasa",
                 "amoreroma",
                 "ABCBACADCBBCDA",
                 "iriwmairanariairie",
                 "anaannaanne",
                 "",
                 "1234432"]



    for test in unit_tests:
        print("{} -> {}".format(test,question2(test)))



if __name__ == '__main__':
    main()


abuttuba -> abuttuba
acaramanamaraca -> acaramanamaraca
adogapanicinapagoda -> adogapanicinapagoda
anutforajaroftuna -> anutforajaroftuna
asantaatnasa -> asantaatnasa
amoreroma -> amoreroma
ABCBACADCBBCDA -> ADCBBCDA
iriwmairanariairie -> airanaria
anaannaanne -> naannaan
 -> 
1234432 -> 234432


## Answer
The complexity of the algorithm above will be O(N^2 / 2) because there is a nested loop and necessarily we have to iterate through all the string everytime. This is the simplest form I found to to make this routine work on the spot, it is still functional and easy to understand. The choice for nested loops was due to be the most logical solution to look within a string for substrings that are palindromes. Since every 2 letter combination can be a palindrome, all possible combinations of letters within the string will have to be tested. For example, the string abcba will have the following breakdown:

* Iteration: 0 -> a
* Iteration: 1 -> ab
* Iteration: 1 -> ab
* Iteration: 2 -> abc
* Iteration: 3 -> abcb
* Iteration: 4 -> abcba
* Iteration: 5 -> b
* Iteration: 6 -> bc
* Iteration: 7 -> bcb
* Iteration: 8 -> bcba
* Iteration: 9 -> c
* Iteration: 10 -> cb
* Iteration: 11 -> cba
* Iteration: 12 -> b
* Iteration: 13 -> ba
* Iteration: 14 -> a

Basically, the Complexity of the Algorithm will be exactly O(1/2*N*(N+1)) where N is the length of the String. This is better than O(N^2) by almost 50%.
Given the test cases bellow:

* Case 1: A But Tuba
* Case 2: A car, a man, a maraca
* Case 3: A dog! A panic in a pagoda
* Case 4: A nut for a jar of tuna
* Case 5: A Santa at Nasa.
* Case 6: Amore, Roma.
* Case 7: ABCBAC ADCBBCDA
* Case 8: iriwm airanaria irie
* Case 9: a naannaan ne
* Case 10: null entry
* Case 11: 1 234432 

All tests are passed as it can be seen above.


# Question 3

Given an undirected graph G, find the minimum spanning tree within G. A minimum spanning tree connects all vertices in a graph with the smallest possible total weight of edges. Your function should take in and return an adjacency list structured like this:

```python
{'A': [('B', 2)],
 'B': [('A', 2), ('C', 5)], 
 'C': [('B', 5)]}

```


Vertices are represented as unique strings. The function definition should be `question3(G)`

In [3]:
# Python Program for union-find algorithm to detect cycle in a undirected graph
from collections import defaultdict
debug = False

#This class represents a undirected graph using adjacency list representation
class Graph:
    def __init__(self,vertices):
        self.V= vertices #No. of vertices
        self.graph = defaultdict(list) # default dictionary to store graph
    # function to add an edge to graph
    def addEdge(self,u,v):
        self.graph[u].append(v)
    # function to remove edge from Graph
    def removeEdge(self,u,v):
        del self.graph[u][-1]

    # A utility function to find the subset of an element i
    def find_parent(self, parent,i):
        if parent[i] == -1:
            return i
        if parent[i]!= -1:
             return self.find_parent(parent,parent[i])

    # A utility function to do union of two subsets
    def union(self,parent,x,y):
        x_set = self.find_parent(parent, x)
        y_set = self.find_parent(parent, y)
        parent[x_set] = y_set

    # The main function to check whether a given graph
    # contains cycle or not
    def isCyclic(self):
        # Allocate memory for creating V subsets and
        # Initialize all subsets as single element sets
        parent = [-1]*(self.V)
        # Iterate through all edges of graph, find subset of both
        # vertices of every edge, if both subsets are same, then
        # there is cycle in graph.
        for i in self.graph:
            for j in self.graph[i]:
                x = self.find_parent(parent, i)
                y = self.find_parent(parent, j)
                if x == y:
                    return True
                self.union(parent,x,y)

    def node_hash(self, keys):
        # Function to transform graphs where Nodes are Chars instead of Ints
        self.str_hash = {}
        node = 0
        for key in keys:
            # if the key is a string and its still not in ditct
            if str(key) == key and self.str_hash.get(key, 'No Key') == 'No Key':
                self.str_hash[key] = node
                node += 1;
            elif int(key) == key: # if key is a Int
                self.str_hash[key] = key


    def reverse_hash(self, value):
        # Function to obtain key name from value in dict
        for key in self.str_hash:
            if self.str_hash[key] == value:
                return key

    def adjacency_list(self):
        # Creates Adjacent List to present results
        self.adj_list = defaultdict(list)
        for min_edge in self.mst:
            self.adj_list[min_edge[0]].append((min_edge[1],min_edge[2]))


    def KruskalMST(self):
    # Kruskal’s Minimum Spanning Tree Algorithm
        self.mst = [] # Minimum Spanning Tree
        for w,s,d in zip(self.weight,self.src,self.dest):
            self.addEdge(s,d) # Add a new edge to graph
            if self.isCyclic(): # If edge added makes a cycle, reject
                self.removeEdge(s,d)
                if debug: print("Graph contains cycle")
            else : # if no cycle, add edge to minimum spanning tree
                self.mst.append([self.reverse_hash(s), self.reverse_hash(d), w])
                if debug: print ("Graph does not contain cycle ")
        return(self.adjacency_list())

    def order_edges(self, graph_adj_list):
        # Initiate Variables
        self.src, self.dest, self.weight = [],[],[]
        for key, content in graph_adj_list.items():
            for entry in content:
                self.src.append(self.str_hash[key])
                self.dest.append(self.str_hash[entry[0]])
                self.weight.append(entry[1])
        # return weight, src, dest sorted by edge weight
        self.weight, self.src, self.dest  = zip(*sorted(zip(self.weight, self.src, self.dest)))


def question3(G):
    g = Graph(len(G)) # Create a graph with given adjacency list
    g.node_hash(G.keys()) # Hash node names, in case they are strigs
    g.order_edges(G) # Order all edges by weight, ascending order
    g.KruskalMST() # Run Kruska Greedy Algorith
    return dict(g.adj_list) # Put together final adjacency List with MST


def main():
    G1 = {0: [(1,4),(7,8)],
        1: [(2,8),(7,11)],
        2: [(3,7),(8,2),(5,4)],
        3: [(4,9),(5,14)],
        4: [(5,10)],
        5: [(6,2)],
        6: [(8,6),(7,1)],
        7: [(8,7)],
        8: []}
    G = {'A': [('B', 2)],
          'B': [('A', 2), ('C', 5)],
          'C': [('B', 5)]}

    print(question3(G))

if __name__ == '__main__':
    main()


{'A': [('B', 2)], 'B': [('C', 5)]}


# Answer
This was the hardest coding question in this set. Returning the Minimun Spanning Tree (MST) given the graph in the form of a Adjency List (AL). I created the `class Graph` to handle all instances realted to a Graph. If we look at the function `question3(G)` we see the order of operations I chose to perform.

`g = Graph(len(G))` First I create a object type Graph, which received the length of G the input. This will be the number of nodes in our Graph.

`g.node_hash(G.keys())` I create this function to handle some complication I had while coding this solution. I needed the nodes to be integers, but the question is handed nodes with `Chars` *A,B,C...* I decided to hash them, so A will be 0, B will 1 for instance. Then I solve the problem and in the end I return my solution in the original form. This is, if it was given Char Nodes they will be returned as such, if they were given integers, then that will be the outcome. 

`g.order_edges(G)` This function will order all edges in ascending weight. This is done so we run the Greedy Algorithm, *Kruskal*

`g.KruskalMST()` After all edges are ordered, a for loop will go through each, connection the graph, making sure each edge addition does not form a cycle within the graph. Once all Nodes are connected, the function will return the outcome.

`g.adj_list` Return a dictionary, same format as it was input with the MST.

This algorithm has the complexity of *O(E log V)* where E are number of edges and V number of Nodes or Vertices. We apply sorting on all Edges and then we loop throgh all of them.


# Question 4

Find the least common ancestor between two nodes on a binary search tree. The least common ancestor is the farthest node from the root that is an ancestor of both nodes. For example, the root is a common ancestor of all nodes on the tree, but if both nodes are descendents of the root's left child, then that left child might be the lowest common ancestor. You can assume that both nodes are in the tree, and the tree itself adheres to all BST properties. The function definition should look like question4(T, r, n1, n2), where T is the tree represented as a matrix, where the index of the list is equal to the integer stored in that node and a 1 represents a child node, r is a non-negative integer representing the root, and n1 and n2 are non-negative integers representing the two nodes in no particular order. For example, one test case might be


In [1]:
class Node(object):
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None

class BST(object):
    def __init__(self, root):
        self.root = Node(root)

    def insert(self, new_val):
        self.insert_helper(self.root, new_val)

    def insert_helper(self, current, new_val):
        if current.value < new_val: # Push sub-node to right
            if current.right:
                self.insert_helper(current.right, new_val)
            else:
                current.right = Node(new_val)
        else: # Push sub-node to left
            if current.left:
                self.insert_helper(current.left, new_val)
            else:
                current.left = Node(new_val)

    def search(self, find_val):
        return self.search_helper(self.root, find_val)

    def search_helper(self, current, find_val):
        if current:
            if current.value == find_val:
                return True
            elif current.value < find_val:
                return self.search_helper(current.right, find_val)
            else:
                return self.search_helper(current.left, find_val)
        return False

    def makeBST(self, matrix):
        for i, subnode in enumerate(matrix[self.root.value]): # Start at root row
            if subnode !=0: # if subnode is a child, insert it.
                self.insert(i)

        for i,node in enumerate(matrix):
            for j,child in enumerate(node):
                if child != 0 and i != self.root.value:
                    self.insert(j)
                    #print("Node {} has child {}".format(i,j))

    def lca(self, head_node, n1, n2):
        # Base Case
        if head_node is None:
            return None

        # # If both n1 and n2 are smaller than root, then LCA
        # # lies in left
        if(head_node.value > n1 and head_node.value > n2):
            return self.lca(head_node.left, n1, n2)

        # If both n1 and n2 are greater than root, then LCA
        # lies in right
        if(head_node.value < n1 and head_node.value < n2):
            return self.lca(head_node.right, n1, n2)

        return head_node.value



def question4(matrix, root, n1, n2):
    head = BST(root)
    head.makeBST(matrix)
    return head.lca(head.root,n1, n2)

# Main program
def main():
    print(question4([[0, 1, 0, 0, 0],
                     [0, 0, 0, 0, 0],
                     [0, 0, 0, 0, 0],
                     [1, 0, 0, 0, 1],
                     [0, 0, 0, 0, 0]],
                     3, 1, 4))

if __name__ == '__main__':
    main()


3


# Answer
Given a Binary Search Tree represented in Matrix form, determining the Least Common Ancestral (LCA) takes a lot of code for what it seems to be a very simple task. Although, the `class BST(object):` has a couple extra functions not needed for this answer, but I included them for the sake of completness. I chose to create two classes, `node` and `BST`. The Node class will just include the vlue of the Node and if It has a subnode to the Right or Left. If this it he casem subnode will be class *Node*. 

`BST` has function to create the Tree, adding nodes accordingly to the input Matrix. This include the function `lca` which will return the Least Common Ancestral, which is the value of the Node that is common to both *n1* and *n2*. This Algorithm takes advantage of the way binary search trees are buit. Where all subnode values larger to the node are pushed to the right and smaller to the left. This makes it very easy to search for values within the tree, making the search time of linear complexity.

Given n1 and n2, it is easy to see if the root node above is the LCA or not. If n1 is larger than root and n2 smaller, this means the root has to be the LCA. Other cases will be if n1 and n2 are both smaller or large, then we can advance on step deeper on the tree, and make a new root with the node at that step. This cycle will continue until n1 is larger than root, and n2 smaller or vice-versa.


# Question 5

Find the element in a singly linked list that's m elements from the end. For example, if a linked list has 5 elements, the 3rd element from the end is the 3rd element. The function definition should look like question5(ll, m), where ll is the first node of a linked list and m is the "mth number from the end". You should copy/paste the Node class below to use as a representation of a node in the linked list. Return the value of the node at that position.

```Python
class Node(object):
  def __init__(self, data):
    self.data = data
    self.next = None
```

In [2]:
class Node(object):
	def __init__(self, data):
		self.data = data
		self.next = None



def question5(ll, m):
	# Setting up initial values - head of linked list
	p1 = ll
	p2 = ll

	# This cycle sets p1 m positions in front of p2
	for i in range(0, m):
		if (p1 == None):
			return None
		p1 = p1.next
	# This while loop will run until p1 get to the last element of list
	# at the time p2 will be m positions behin
	# p2 value is returned, since that with be the mth element
	# in the linked list
	while (p1 != None):
		p1 = p1.next
		p2 = p2.next
	return p2.data


A = Node(2)
B = Node(3)
C = Node(8)
D = Node(1)
E = Node(6)

A.next = B
B.next = C
C.next = D
D.next = E

print(question5(A, 3))


8


# Answer

Given a linked list, find the value of the element mth positions from the end of list. The problem with this situation is that on a linked list, we do not know the total length of the list to start with. Linked lists are basically a vlue with  pointer to the next value. So we do not know all values on the list, just the next one, given any value. This is done to save time and memory space. In comparison, if we just store  whole list, we can always aaccess the mth element from the end using `list[-mth]` making this exercise rather trivial. 

The algorithm for this problem is rather simple. `Class Node(obect)` basically creates a node of the list with a pointer to the next element on the list, if there is one. So creating a linked list such as:
```python
A = Node(2)
B = Node(3)
C = Node(8)
D = Node(1)
E = Node(6)

A.next = B
B.next = C
C.next = D
D.next = E
```

Where basically we create node `A` with value `2` and pointing to next Node `B` with value `3`. So fouth and so on, until the whole Linked List (LL) is created.

Th algorithm basically operates within two loops. First one, `for i in range(0, m):`, will place `p1`, basically a pointer, `m` positions ahead of `p2`, which points to the first node. Then we have `while (p1 != None):` which will cycle through the whole linked list, until it hits the last element, where `p1.next` is none. At that point, `p2`, will be `m` positions lagging behind `p1`, or better saying, `mth` nodes from the end of list. The value of the `mth` node is then returned.

